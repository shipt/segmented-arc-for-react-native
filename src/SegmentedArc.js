import React, { useState, useEffect, useRef, createContext } from 'react';
import { Animated, Easing, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import Svg from 'react-native-svg';
import Segment from './components/Segment';
import Cap from './components/Cap';
import RangesDisplay from './components/RangesDisplay';

const SegmentedArcContext = createContext();

export const SegmentedArc = ({
  fillValue = 0,
  segments = [],
  filledArcWidth = 8,
  emptyArcWidth = 8,
  spaceBetweenSegments = 2,
  arcDegree = 180,
  radius = 100,
  animationDuration = 1000,
  isAnimated = true,
  animationDelay = 0,
  showArcRanges = false,
  middleContentContainerStyle = {},
  ranges = [],
  rangesTextColor = '#000000',
  rangesTextStyle = styles.rangeTextStyle,
  capInnerColor = '#28E037',
  capOuterColor = '#FFFFFF',
  alignRangesWithSegments = true,
  children
}) => {
  const [arcAnimatedValue] = useState(new Animated.Value(0));
  const animationRunning = useRef(false);

  if (segments.length === 0) {
    return null;
  }

  const totalArcs = segments.length;
  const totalSpaces = totalArcs - 1;
  const totalSpacing = totalSpaces * spaceBetweenSegments;

  const arcSegmentDegree = (arcDegree - totalSpacing) / totalArcs;
  const arcsStart = 90 - arcDegree / 2;

  const effectiveRadius = radius + Math.max(filledArcWidth, emptyArcWidth);
  const margin = 12;
  let svgHeight = effectiveRadius + margin * 1.5;
  const svgWidth = effectiveRadius * 2 + 2 * margin;
  const center = effectiveRadius + margin;

  if (arcDegree > 180) {
    const offsetAngle = (arcDegree - 180) / 2;
    const heightOffset = effectiveRadius * Math.sin(offsetAngle);
    svgHeight += heightOffset + 2 * margin + filledArcWidth;
  }

  const localMiddleContentContainerStyle = {
    position: 'absolute',
    alignSelf: 'center',
    top: filledArcWidth + margin + 16,
    width: radius + filledArcWidth * 2,
    height: radius,
    alignItems: 'center',
    overflow: 'hidden',
    ...middleContentContainerStyle
  };

  const _ensureDefaultSegmentScale = () => {
    const segmentsWithoutScale = segments.filter(segment => !segment.scale);
    const allocatedScale = segments.reduce((total, current) => total + (current.scale || 0), 0);
    const defaultArcScale = (1 - allocatedScale) / segmentsWithoutScale.length;
    segmentsWithoutScale.forEach(segment => (segment.scale = defaultArcScale));
  };

  const _ensureDefaultArcScale = () => {
    const segmentsWithoutArcDegreeScaleLength = segments.filter(
      segment => typeof segment.arcDegreeScale !== 'number'
    ).length;
    const totalProvidedArcDegreeScale = segments.reduce((acc, segment) => acc + (segment.arcDegreeScale ?? 0), 0);
    segments.forEach(segment => {
      if (typeof segment.arcDegreeScale !== 'number')
        segment.arcDegreeScale = (1 - totalProvidedArcDegreeScale) / segmentsWithoutArcDegreeScaleLength;
    });
  };

  let remainingValue = fillValue;

  _ensureDefaultSegmentScale();
  _ensureDefaultArcScale();

  let arcs = [];
  segments.forEach((segment, index) => {
    const arcDegreeScale = segment.arcDegreeScale;
    const previousSegmentEnd = !!index ? arcs[index - 1].end : arcsStart;
    const start = previousSegmentEnd + (!!index ? spaceBetweenSegments : 0);
    const end = (arcDegree - totalSpacing) * arcDegreeScale + start;

    const valueMax = 100 * segment.scale;
    const effectiveScaledValue = Math.min(remainingValue, valueMax);
    const scaledPercentage = effectiveScaledValue / valueMax;
    const filled = start + scaledPercentage * (end - start);
    remainingValue -= effectiveScaledValue;

    const newArc = {
      centerX: center,
      centerY: center,
      start,
      end,
      filled,
      isComplete: filled === end,
      filledColor: segment.filledColor,
      emptyColor: segment.emptyColor,
      data: segment.data
    };

    arcs.push(newArc);
  });

  const lastFilledSegment = arcs.find(a => a.filled !== a.end) || arcs[arcs.length - 1];

  useEffect(() => {
    if (!lastFilledSegment) return;
    if (animationRunning.current) return;
    if (!isAnimated) return;
    animationRunning.current = true;
    Animated.timing(arcAnimatedValue, {
      toValue: lastFilledSegment.filled,
      duration: animationDuration,
      delay: animationDelay,
      useNativeDriver: false,
      easing: Easing.out(Easing.ease)
    }).start();

    const listenerId = arcAnimatedValue.addListener(e => {
      if (e.value === lastFilledSegment.filled) animationRunning.current = false;
    });
    return () => arcAnimatedValue.removeListener(listenerId);
  }, []);

  if (arcs.length === 0) {
    return null;
  }

  return (
    <View style={styles.container} testID="container">
      <Svg width={svgWidth} height={svgHeight}>
        <SegmentedArcContext.Provider
          value={{
            arcs,
            margin,
            center,
            filledArcWidth,
            radius,
            isAnimated,
            animationDuration,
            emptyArcWidth,
            totalArcs,
            arcsStart,
            spaceBetweenSegments,
            arcSegmentDegree,
            arcAnimatedValue,
            lastFilledSegment,
            ranges,
            rangesTextColor,
            rangesTextStyle,
            capInnerColor,
            capOuterColor,
            alignRangesWithSegments
          }}
        >
          {arcs.map((arc, index) => (
            <Segment key={index.toString()} arc={arc} />
          ))}

          <Cap />

          {showArcRanges && <RangesDisplay />}
        </SegmentedArcContext.Provider>
      </Svg>

      {children && <View style={localMiddleContentContainerStyle}>{children({ lastFilledSegment })}</View>}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center'
  },
  rangeTextStyle: {
    fontSize: 12
  }
});

SegmentedArc.propTypes = {
  fillValue: PropTypes.number,
  segments: PropTypes.arrayOf(
    PropTypes.shape({
      scale: PropTypes.number,
      filledColor: PropTypes.string.isRequired,
      emptyColor: PropTypes.string.isRequired,
      data: PropTypes.object,
      arcDegreeScale: PropTypes.number
    })
  ).isRequired,
  filledArcWidth: PropTypes.number,
  emptyArcWidth: PropTypes.number,
  spaceBetweenSegments: PropTypes.number,
  arcDegree: PropTypes.number,
  radius: PropTypes.number,
  animationDuration: PropTypes.number,
  isAnimated: PropTypes.bool,
  animationDelay: PropTypes.number,
  showArcRanges: PropTypes.bool,
  children: PropTypes.func,
  middleContentContainerStyle: PropTypes.object,
  ranges: PropTypes.arrayOf(PropTypes.string),
  rangesTextColor: PropTypes.string,
  rangesTextStyle: PropTypes.object,
  capInnerColor: PropTypes.string,
  capOuterColor: PropTypes.string,
  alignRangesWithSegments: PropTypes.bool
};
export { SegmentedArcContext };
export default SegmentedArc;
