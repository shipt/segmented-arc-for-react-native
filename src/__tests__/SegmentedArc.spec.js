import React from 'react';
import { Text, Animated, Easing } from 'react-native';
import { SegmentedArc } from '../SegmentedArc';
import { render } from '@testing-library/react-native';

describe('SegmentedArc', () => {
  let segments = [
    {
      scale: 0.25,
      color: '#FF0000',
      label: 'First Segment'
    },
    {
      scale: 0.02,
      color: '#FFA500',
      label: 'Second Segment'
    },
    {
      scale: 0.02,
      color: '#00FF00',
      label: 'Third Segment'
    },
    {
      scale: 0.02,
      color: '#0000FF',
      label: 'Fourth Segment'
    }
  ];

  const testId = 'container';
  let props = {};
  let wrapper;

  const getWrapper = properties => {
    return render(<SegmentedArc {...properties} />);
  };

  beforeEach(() => {
    Animated.timing = jest.fn();
    Easing.out = jest.fn();
    Easing.ease = jest.fn();
    Animated.timing.mockReturnValue({ start: jest.fn() });
    props = { segments, fillValue: 50 };
  });

  afterEach(() => {
    Animated.timing.mockReset();
    Easing.out.mockReset();
    Easing.ease.mockReset();
  });

  it('renders default', () => {
    wrapper = getWrapper(props);
    expect(wrapper.getByTestId(testId).props).toMatchSnapshot();
    expect(Animated.timing).toHaveBeenCalledTimes(1);
    expect(Easing.out).toHaveBeenCalledWith(Easing.ease);
  });

  it('does not render the component when segments is not provided', () => {
    wrapper = getWrapper({ ...props, segments: [] });
    expect(wrapper.queryByTestId(testId)).toBeNull();
  });

  it("automatically increases the component's height when totalArcSize is greater than 180 degrees", () => {
    wrapper = getWrapper({ ...props, totalArcSize: 270 });
    expect(wrapper.getByTestId(testId).props).toMatchSnapshot();
  });

  it('renders arc with ranges', () => {
    props.showArcRanges = true;
    props.ranges = ['1.25', '2.5', '3.75', '5'];
    wrapper = getWrapper(props);
    expect(wrapper.getByTestId(testId).props).toMatchSnapshot('shows ranges');
  });

  it('renders with middle content', () => {
    wrapper = render(
      <SegmentedArc {...props}>{lastFilledSegment => <Text>{lastFilledSegment.label}</Text>}</SegmentedArc>
    );
    expect(wrapper.getByTestId(testId).props).toMatchSnapshot();
  });

  it('does not define animation', () => {
    props.isAnimated = false;
    wrapper = getWrapper(props);
    expect(Animated.timing).not.toHaveBeenCalled();
    expect(Easing.out).not.toHaveBeenCalledWith(Easing.ease);
  });

  it('sets the last segment for lastFilledSegment prop when fillValue is equal or greater than 100%', () => {
    props.fillValue = 100;
    wrapper = getWrapper(props);
    expect(wrapper.getByTestId(testId).props).toMatchSnapshot();
  });
});
