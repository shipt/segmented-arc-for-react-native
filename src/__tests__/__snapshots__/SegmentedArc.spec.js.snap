// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedArc automatically increases the component's height when arcDegree is greater than 180 degrees 1`] = `
{
  "children": [
    <Svg
      height={249.8975806496848}
      preserveAspectRatio="xMidYMid meet"
      width={240}
    >
      <Context.Provider
        value={
          {
            "animationDuration": 1000,
            "arcAnimatedValue": 0,
            "arcSegmentDegree": 66,
            "arcsStart": -45,
            "capInnerColor": "#28E037",
            "capOuterColor": "#FFFFFF",
            "center": 120,
            "emptyArcWidth": 8,
            "filledArcWidth": 8,
            "isAnimated": true,
            "lastFilledSegment": {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 225,
              "filled": 225,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 159,
            },
            "margin": 12,
            "radius": 100,
            "ranges": [],
            "rangesTextColor": "#000000",
            "rangesTextStyle": {
              "fontSize": 12,
            },
            "spaceBetweenSegments": 2,
            "totalArcs": 4,
          }
        }
      >
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "First Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 21,
              "filled": 21,
              "filledColor": "#FF0000",
              "isComplete": true,
              "start": -45,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Second Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 89,
              "filled": 89,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 23,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 157,
              "filled": 157,
              "filledColor": "#00FF00",
              "isComplete": true,
              "start": 91,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 225,
              "filled": 225,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 159,
            }
          }
        />
        <Cap />
      </Context.Provider>
    </Svg>,
    undefined,
  ],
  "style": {
    "alignItems": "center",
  },
  "testID": "container",
}
`;

exports[`SegmentedArc renders arc with ranges: shows ranges 1`] = `
{
  "children": [
    <Svg
      height={126}
      preserveAspectRatio="xMidYMid meet"
      width={240}
    >
      <Context.Provider
        value={
          {
            "animationDuration": 1000,
            "arcAnimatedValue": 0,
            "arcSegmentDegree": 43.5,
            "arcsStart": 0,
            "capInnerColor": "#28E037",
            "capOuterColor": "#FFFFFF",
            "center": 120,
            "emptyArcWidth": 8,
            "filledArcWidth": 8,
            "isAnimated": true,
            "lastFilledSegment": {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            },
            "margin": 12,
            "radius": 100,
            "ranges": [
              "1.25",
              "2.5",
              "3.75",
              "5",
            ],
            "rangesTextColor": "#000000",
            "rangesTextStyle": {
              "fontSize": 12,
            },
            "spaceBetweenSegments": 2,
            "totalArcs": 4,
          }
        }
      >
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "First Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 43.5,
              "filled": 43.5,
              "filledColor": "#FF0000",
              "isComplete": true,
              "start": 0,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Second Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 89,
              "filled": 89,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 45.5,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 134.5,
              "filled": 134.5,
              "filledColor": "#00FF00",
              "isComplete": true,
              "start": 91,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            }
          }
        />
        <Cap />
        <RangesDisplay />
      </Context.Provider>
    </Svg>,
    undefined,
  ],
  "style": {
    "alignItems": "center",
  },
  "testID": "container",
}
`;

exports[`SegmentedArc renders default 1`] = `
{
  "children": [
    <Svg
      height={126}
      preserveAspectRatio="xMidYMid meet"
      width={240}
    >
      <Context.Provider
        value={
          {
            "animationDuration": 1000,
            "arcAnimatedValue": 0,
            "arcSegmentDegree": 43.5,
            "arcsStart": 0,
            "capInnerColor": "#28E037",
            "capOuterColor": "#FFFFFF",
            "center": 120,
            "emptyArcWidth": 8,
            "filledArcWidth": 8,
            "isAnimated": true,
            "lastFilledSegment": {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            },
            "margin": 12,
            "radius": 100,
            "ranges": [],
            "rangesTextColor": "#000000",
            "rangesTextStyle": {
              "fontSize": 12,
            },
            "spaceBetweenSegments": 2,
            "totalArcs": 4,
          }
        }
      >
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "First Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 43.5,
              "filled": 43.5,
              "filledColor": "#FF0000",
              "isComplete": true,
              "start": 0,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Second Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 89,
              "filled": 89,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 45.5,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 134.5,
              "filled": 134.5,
              "filledColor": "#00FF00",
              "isComplete": true,
              "start": 91,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            }
          }
        />
        <Cap />
      </Context.Provider>
    </Svg>,
    undefined,
  ],
  "style": {
    "alignItems": "center",
  },
  "testID": "container",
}
`;

exports[`SegmentedArc renders segments with arc degree scales 1`] = `
{
  "children": [
    <Svg
      height={126}
      preserveAspectRatio="xMidYMid meet"
      width={240}
    >
      <Context.Provider
        value={
          {
            "animationDuration": 1000,
            "arcAnimatedValue": 0,
            "arcSegmentDegree": 58.666666666666664,
            "arcsStart": 0,
            "capInnerColor": "#28E037",
            "capOuterColor": "#FFFFFF",
            "center": 120,
            "emptyArcWidth": 8,
            "filledArcWidth": 8,
            "isAnimated": true,
            "lastFilledSegment": {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 144.8,
            },
            "margin": 12,
            "radius": 100,
            "ranges": [],
            "rangesTextColor": "#000000",
            "rangesTextStyle": {
              "fontSize": 12,
            },
            "spaceBetweenSegments": 2,
            "totalArcs": 3,
          }
        }
      >
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "First Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 88,
              "filled": 88,
              "filledColor": "#FF0000",
              "isComplete": true,
              "start": 0,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Second Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 142.8,
              "filled": 142.8,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 90,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 144.8,
            }
          }
        />
        <Cap />
      </Context.Provider>
    </Svg>,
    undefined,
  ],
  "style": {
    "alignItems": "center",
  },
  "testID": "container",
}
`;

exports[`SegmentedArc renders with ensured arc degree scales when missing from segments 1`] = `
{
  "children": [
    <Svg
      height={126}
      preserveAspectRatio="xMidYMid meet"
      width={240}
    >
      <Context.Provider
        value={
          {
            "animationDuration": 1000,
            "arcAnimatedValue": 0,
            "arcSegmentDegree": 58.666666666666664,
            "arcsStart": 0,
            "capInnerColor": "#28E037",
            "capOuterColor": "#FFFFFF",
            "center": 120,
            "emptyArcWidth": 8,
            "filledArcWidth": 8,
            "isAnimated": true,
            "lastFilledSegment": {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 136,
            },
            "margin": 12,
            "radius": 100,
            "ranges": [],
            "rangesTextColor": "#000000",
            "rangesTextStyle": {
              "fontSize": 12,
            },
            "spaceBetweenSegments": 2,
            "totalArcs": 3,
          }
        }
      >
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "First Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 88,
              "filled": 88,
              "filledColor": "#FF0000",
              "isComplete": true,
              "start": 0,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Second Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 134,
              "filled": 134,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 90,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 136,
            }
          }
        />
        <Cap />
      </Context.Provider>
    </Svg>,
    undefined,
  ],
  "style": {
    "alignItems": "center",
  },
  "testID": "container",
}
`;

exports[`SegmentedArc renders with middle content 1`] = `
{
  "children": [
    <Svg
      height={126}
      preserveAspectRatio="xMidYMid meet"
      width={240}
    >
      <Context.Provider
        value={
          {
            "animationDuration": 1000,
            "arcAnimatedValue": 0,
            "arcSegmentDegree": 43.5,
            "arcsStart": 0,
            "capInnerColor": "#28E037",
            "capOuterColor": "#FFFFFF",
            "center": 120,
            "emptyArcWidth": 8,
            "filledArcWidth": 8,
            "isAnimated": true,
            "lastFilledSegment": {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            },
            "margin": 12,
            "radius": 100,
            "ranges": [],
            "rangesTextColor": "#000000",
            "rangesTextStyle": {
              "fontSize": 12,
            },
            "spaceBetweenSegments": 2,
            "totalArcs": 4,
          }
        }
      >
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "First Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 43.5,
              "filled": 43.5,
              "filledColor": "#FF0000",
              "isComplete": true,
              "start": 0,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Second Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 89,
              "filled": 89,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 45.5,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 134.5,
              "filled": 134.5,
              "filledColor": "#00FF00",
              "isComplete": true,
              "start": 91,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            }
          }
        />
        <Cap />
      </Context.Provider>
    </Svg>,
    <View
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "height": 100,
          "overflow": "hidden",
          "position": "absolute",
          "top": 36,
          "width": 116,
        }
      }
    >
      <Text />
    </View>,
  ],
  "style": {
    "alignItems": "center",
  },
  "testID": "container",
}
`;

exports[`SegmentedArc sets the last segment for lastFilledSegment prop when fillValue is equal or greater than 100% 1`] = `
{
  "children": [
    <Svg
      height={126}
      preserveAspectRatio="xMidYMid meet"
      width={240}
    >
      <Context.Provider
        value={
          {
            "animationDuration": 1000,
            "arcAnimatedValue": 0,
            "arcSegmentDegree": 43.5,
            "arcsStart": 0,
            "capInnerColor": "#28E037",
            "capOuterColor": "#FFFFFF",
            "center": 120,
            "emptyArcWidth": 8,
            "filledArcWidth": 8,
            "isAnimated": true,
            "lastFilledSegment": {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            },
            "margin": 12,
            "radius": 100,
            "ranges": [],
            "rangesTextColor": "#000000",
            "rangesTextStyle": {
              "fontSize": 12,
            },
            "spaceBetweenSegments": 2,
            "totalArcs": 4,
          }
        }
      >
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "First Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 43.5,
              "filled": 43.5,
              "filledColor": "#FF0000",
              "isComplete": true,
              "start": 0,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Second Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 89,
              "filled": 89,
              "filledColor": "#FFA500",
              "isComplete": true,
              "start": 45.5,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Third Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 134.5,
              "filled": 134.5,
              "filledColor": "#00FF00",
              "isComplete": true,
              "start": 91,
            }
          }
        />
        <Segment
          arc={
            {
              "centerX": 120,
              "centerY": 120,
              "data": {
                "label": "Fourth Segment",
              },
              "emptyColor": "#F2F3F5",
              "end": 180,
              "filled": 180,
              "filledColor": "#0000FF",
              "isComplete": true,
              "start": 136.5,
            }
          }
        />
        <Cap />
      </Context.Provider>
    </Svg>,
    undefined,
  ],
  "style": {
    "alignItems": "center",
  },
  "testID": "container",
}
`;
